### PR TITLE
HBASE-23862 Fix flaky TestSnapshotFromMaster in 1.x versions

### PR DESCRIPTION
**Pull request by jatsakthi**
Backport of HBASE-23658 - Fix flaky TestSnapshotFromMaster
Signed-off-by: Duo Zhang <zhangduo@apache.org>
(cherry picked from commit e750d2c7af03bfcb85422077cf59ab4b1815cb12)

### CODE REVIEW

File: hbase-server/src/test/java/org/apache/hadoop/hbase/master/cleaner/TestSnapshotFromMaster.java [hunk 1]
```diff
@@ -530,8 +530,11 @@ public boolean evaluate() throws Exception {
         return UTIL.getHBaseAdmin().listSnapshots(Pattern.compile(snapshotName)).size() == 1;
       }
     });
-    assertTrue(master.getSnapshotManager().isTakingAnySnapshot());
-    Thread.sleep(11 * 1000L);
-    assertFalse(master.getSnapshotManager().isTakingAnySnapshot());
+    UTIL.waitFor(30000, new Predicate<Exception>() {
+      @Override
+      public boolean evaluate() throws Exception {
+        return UTIL.getHBaseAdmin().listSnapshots(Pattern.compile(snapshotName)).size() == 1;
```
Comment (infraio) [inline hbase-server/src/test/java/org/apache/hadoop/hbase/master/cleaner/TestSnapshotFromMaster.java]: UTIL.waitFor(30000, () -> !master.getSnapshotManager().isTakingAnySnapshot());?

**Comment by jatsakthi** [comment]
Thanks for pointing out the predicate, @infraio

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   7m 17s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ branch-1 Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   8m 25s |  branch-1 passed  |
| +1 :green_heart: |  compile  |   0m 42s |  branch-1 passed with JDK v1.8.0_242  |
| +1 :green_heart: |  compile  |   0m 43s |  branch-1 passed with JDK v1.7.0_252  |
| +1 :green_heart: |  checkstyle  |   1m 36s |  branch-1 passed  |
| +1 :green_heart: |  shadedjars  |   2m 58s |  branch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 40s |  branch-1 passed with JDK v1.8.0_242  |
| +1 :green_heart: |  javadoc  |   0m 40s |  branch-1 passed with JDK v1.7.0_252  |
| +0 :ok: |  spotbugs  |   2m 50s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   2m 46s |  branch-1 passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   1m 54s |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 41s |  the patch passed with JDK v1.8.0_242  |
| +1 :green_heart: |  javac  |   0m 41s |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 44s |  the patch passed with JDK v1.7.0_252  |
| +1 :green_heart: |  javac  |   0m 44s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   1m 28s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedjars  |   2m 45s |  patch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  hadoopcheck  |   4m 53s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2.  |
| +1 :green_heart: |  javadoc  |   0m 32s |  the patch passed with JDK v1.8.0_242  |
| +1 :green_heart: |  javadoc  |   0m 39s |  the patch passed with JDK v1.7.0_252  |
| +1 :green_heart: |  findbugs  |   2m 46s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  | 128m 31s |  hbase-server in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 34s |  The patch does not generate ASF License warnings.  |
|  |   | 174m 49s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.6 Server=19.03.6 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1181/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/1181 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux cc0d232e09e2 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/Base-PreCommit-GitHub-PR_PR-1181/out/precommit/personality/provided.sh |
| git revision | branch-1 / b9c676c |
| Default Java | 1.7.0_252 |
| Multi-JDK versions | /usr/lib/jvm/zulu-8-amd64:1.8.0_242 /usr/lib/jvm/zulu-7-amd64:1.7.0_252 |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1181/1/testReport/ |
| Max. process+thread count | 4587 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1181/1/console |
| versions | git=1.9.1 maven=3.0.5 findbugs=3.0.1 |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 42s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ branch-1 Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   8m 23s |  branch-1 passed  |
| +1 :green_heart: |  compile  |   0m 41s |  branch-1 passed with JDK v1.8.0_242  |
| +1 :green_heart: |  compile  |   0m 43s |  branch-1 passed with JDK v1.7.0_252  |
| +1 :green_heart: |  checkstyle  |   1m 37s |  branch-1 passed  |
| +1 :green_heart: |  shadedjars  |   2m 54s |  branch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 40s |  branch-1 passed with JDK v1.8.0_242  |
| +1 :green_heart: |  javadoc  |   0m 40s |  branch-1 passed with JDK v1.7.0_252  |
| +0 :ok: |  spotbugs  |   2m 52s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   2m 49s |  branch-1 passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   1m 56s |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 40s |  the patch passed with JDK v1.8.0_242  |
| +1 :green_heart: |  javac  |   0m 40s |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 44s |  the patch passed with JDK v1.7.0_252  |
| +1 :green_heart: |  javac  |   0m 44s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   1m 32s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedjars  |   2m 49s |  patch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  hadoopcheck  |   4m 46s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2.  |
| +1 :green_heart: |  javadoc  |   0m 31s |  the patch passed with JDK v1.8.0_242  |
| +1 :green_heart: |  javadoc  |   0m 41s |  the patch passed with JDK v1.7.0_252  |
| +1 :green_heart: |  findbugs  |   2m 50s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  | 126m 39s |  hbase-server in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 34s |  The patch does not generate ASF License warnings.  |
|  |   | 166m 15s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.6 Server=19.03.6 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1181/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/1181 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 357da37bfcc0 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/Base-PreCommit-GitHub-PR_PR-1181/out/precommit/personality/provided.sh |
| git revision | branch-1 / b9c676c |
| Default Java | 1.7.0_252 |
| Multi-JDK versions | /usr/lib/jvm/zulu-8-amd64:1.8.0_242 /usr/lib/jvm/zulu-7-amd64:1.7.0_252 |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1181/2/testReport/ |
| Max. process+thread count | 4795 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1181/2/console |
| versions | git=1.9.1 maven=3.0.5 findbugs=3.0.1 |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

**Comment by jatsakthi** [comment]
@infraio is this one good to go?

**Comment by jatsakthi** [comment]
@busbey this one blocks the release of 1.4.13. Can I get a quick review here, please?

File: hbase-server/src/test/java/org/apache/hadoop/hbase/master/cleaner/TestSnapshotFromMaster.java [hunk 2]
```diff
@@ -530,8 +530,11 @@ public boolean evaluate() throws Exception {
         return UTIL.getHBaseAdmin().listSnapshots(Pattern.compile(snapshotName)).size() == 1;
       }
     });
-    assertTrue(master.getSnapshotManager().isTakingAnySnapshot());
-    Thread.sleep(11 * 1000L);
-    assertFalse(master.getSnapshotManager().isTakingAnySnapshot());
+    UTIL.waitFor(30000, new Predicate<Exception>() {
+      @Override
+      public boolean evaluate() throws Exception {
+        return master.getSnapshotManager().isTakingAnySnapshot();
```
Comment (infraio) [inline hbase-server/src/test/java/org/apache/hadoop/hbase/master/cleaner/TestSnapshotFromMaster.java]: Should add a "!" here? It is assertFalse before.

Comment (jatsakthi) [reply hbase-server/src/test/java/org/apache/hadoop/hbase/master/cleaner/TestSnapshotFromMaster.java]: Oh right! My bad. Overlooked the condition.

Comment (jatsakthi) [reply hbase-server/src/test/java/org/apache/hadoop/hbase/master/cleaner/TestSnapshotFromMaster.java]: Corrected.

**Comment by jatsakthi** [comment]
Thanks @infraio let me push this in to 1.x

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 43s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ branch-1 Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   8m 18s |  branch-1 passed  |
| +1 :green_heart: |  compile  |   0m 40s |  branch-1 passed with JDK v1.8.0_242  |
| +1 :green_heart: |  compile  |   0m 43s |  branch-1 passed with JDK v1.7.0_252  |
| +1 :green_heart: |  checkstyle  |   1m 35s |  branch-1 passed  |
| +1 :green_heart: |  shadedjars  |   2m 58s |  branch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 40s |  branch-1 passed with JDK v1.8.0_242  |
| +1 :green_heart: |  javadoc  |   0m 39s |  branch-1 passed with JDK v1.7.0_252  |
| +0 :ok: |  spotbugs  |   2m 59s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   2m 56s |  branch-1 passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   2m  3s |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 51s |  the patch passed with JDK v1.8.0_242  |
| +1 :green_heart: |  javac  |   0m 51s |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 46s |  the patch passed with JDK v1.7.0_252  |
| +1 :green_heart: |  javac  |   0m 46s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   1m 48s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedjars  |   3m 24s |  patch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  hadoopcheck  |   6m 57s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2.  |
| +1 :green_heart: |  javadoc  |   0m 49s |  the patch passed with JDK v1.8.0_242  |
| +1 :green_heart: |  javadoc  |   1m  2s |  the patch passed with JDK v1.7.0_252  |
| +1 :green_heart: |  findbugs  |   4m 23s |  the patch passed  |
||| _ Other Tests _ |
| -1 :x: |  unit  |  44m  8s |  hbase-server in the patch failed.  |
| +1 :green_heart: |  asflicense  |   0m 28s |  The patch does not generate ASF License warnings.  |
|  |   |  90m 21s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.http.log.TestLogLevel |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.6 Server=19.03.6 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1181/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/1181 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 2fbdf925f646 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/Base-PreCommit-GitHub-PR_PR-1181/out/precommit/personality/provided.sh |
| git revision | branch-1 / b9c676c |
| Default Java | 1.7.0_252 |
| Multi-JDK versions | /usr/lib/jvm/zulu-8-amd64:1.8.0_242 /usr/lib/jvm/zulu-7-amd64:1.7.0_252 |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1181/3/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1181/3/testReport/ |
| Max. process+thread count | 729 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1181/3/console |
| versions | git=1.9.1 maven=3.0.5 findbugs=3.0.1 |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

**Pull request merged**